File: Sliding Window/LongestSubstringWithoutRepeatingCharacters.java


public int lengthOfLongestSubstring(String s) {
    if (s.length() == 0) return 0;
    int end = 0;
    int begin = 0;
    int maxLength = 0;
    HashSet<Character> map = new HashSet<>();

    while (end < s.length()) {
        char character = s.charAt(end);
        if (map.contains(character)) {
            while (map.contains(character)) {
                char beginChar = s.charAt(begin);
                map.remove(beginChar);
                begin++;
            }
        } else {
            map.add(character);
            int len = end - begin + 1;
            if (len > maxLength) {
                maxLength = len;
            }
            end++;
        }
    }

    return maxLength;
}

File: Sliding Window/LongestSubstringWithoutRepeatingCharacters.js
// 3. Longest Substring Without Repeating Characters

// Time Complexity  O(n)
// Space Complexity O(n)

var lengthOfLongestSubstring = function (s) {
    if (!s.length) return 0;
    let end = 0;
    let begin = 0;
    let maxLength = 0;
    const map = new Map();

    while (end < s.length) {
        let char = s[end];
        if (map.has(char)) {
            while (map.has(char)) {
                const beginChar = s[begin];
                map.delete(beginChar);
                begin++;
            }
        } else {
            map.set(char, 1);
            const len = end - begin + 1;
            if (len > maxLength) {
                maxLength = len;
            }
            end++;
        }
    }

    return maxLength;
};
